doc/man/rosie.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "ROSIE" "1" "April 2018" "The Rosie Project" ""
.
.SH "NAME"
\fBrosie\fR \- Rosie Pattern Engine (Rosie Pattern Language)
.
.SH "SYNOPSIS"
\fBrosie\fR [global\-options] command [command\-options] pattern\-expression [file1 [file2 \.\.\.]]
.
.SH "DESCRIPTION"
\fBRosie\fR searches all given input files for lines that match a pattern\. The pattern language is similar to regular expressions, but more powerful\. Also, a set of predefined (named) patterns are provided\. The pattern library is extensible\. Rosie patterns are written in \fBRosie Pattern Language (RPL)\fR\.
.
.P
A shared library, \fBlibrosie\fR, provides a programmatic interface to rosie in C, Go, Python, and other languages that support \fBlibffi\fR\. \fBRosie\fR can be used for data mining on large data sets, or for the kind of smaller tasks that Unix \fBgrep\fR and other regex tools are used for\. The output can be plain text, like \fBgrep\fR produces, or can be structured JSON, in the form of a parse tree\. A number of other output options are available (see below)\.
.
.SH "LANGUAGE REFERENCE"
The RPL Language Reference may be found via the Rosie home page \fIhttp://rosie\-lang\.org\fR (news, docs, and more)\.
.
.SH "COMMANDS"
The primary commands are \fBmatch\fR and \fBgrep\fR\. Help for a command is obtained using the \fB\-h\fR option (see below)\. General help is available with the \fBhelp\fR command\.
.
.TP
\fBconfig\fR
Print the configuration of the Rosie installation, including build information if available\.
.
.TP
\fBexpand\fR \fIexp\fR
Expands a \fBpattern expression\fR entered on the command line\. Output is: (1) the expression as entered (which will show any shell expansions or substitutions); (2) the parsed expression; (3) what the expression means at top level, where a bare expression is a tokenized expression; and (4) the result after any macro expansions\.
.
.TP
\fBgrep\fR \fIexp\fR [file1 \.\.\.]
In the style of grep, match the given \fBpattern\fR against lines in the given input files (or the standard input)\. In this mode, the pattern may match anywhere within a line, and all matches within a line are found\. The default output form in this mode is to print the entire line for all lines that match\.
.
.TP
\fBhelp\fR
Print help text\.
.
.TP
\fBlist\fR [name[\.name]]
List the available patterns from the library whose name matches \fBname\fR\. To list all the names in an imported package, use \fBlist pkgname\.*\fR\.
.
.TP
\fBmatch\fR \fIexp\fR [file1 \.\.\.]
Match the given \fBpattern\fR against lines in the given input files (or the standard input)\. Patterns in RPL begin at the start of the input line\. See the \fBgrep\fR command for behavior that is more like that utility\.
.
.TP
\fBrepl\fR
Enter the read\-eval\-print loop for interactive development and testing of patterns
.
.TP
\fBtest\fR \fIfile1\fR [file2\.\.\.]
Execute the unit tests embedded within the listed rpl files\.
.
.TP
\fBtrace\fR \fIexp\fR [file1 \.\.\.]
Perform a match, but generate a trace listing of all the steps performed in the matching process\. This can be quite a lot of output\.
.
.TP
\fBversion\fR
Print the Rosie version and exit\.
.
.SH "GLOBAL OPTIONS"
.
.TP
\fB\-\-colors\fR \fIcolorspecs\fR
Specify a set of colors and font attributes, each associated with a pattern name\. The format of \fBcolorspecs\fR is a colon\-separated list of \fBname=spec\fR, where:
.
.IP
\fBspec\fR is a semi\-colon separated list of ANSI (SGR) color and font attributes\. Colors may be specified by name or integer code\. Attributes include bold, underline, and others\.
.
.IP
\fBname\fR can be \fB*\fR for a global default; \fBpkgname\.*\fR for a package default; \fBlocalname\fR to match a name exactly, without a prefix; or \fBpkgname\.localname\fR to match an imported name exactly\.
.
.TP
\fB\-f, \-\-file\fR \fIfile\fR
Load a file of rpl code\. This option may be repeated\.
.
.TP
\fB\-\-libpath\fR \fIpaths\fR
Set the rosie \fBlibpath\fR, which is a colon\-separated list of directories to search, in order, for imported packages\. The value is a colon\-separated string\. When the \fBlibpath\fR is not set by the user, Rosie looks for imported packages in the installation directory, at \fBROSIE_LIBDIR\fR (the value of which can be seen using the \fBrosie config\fR command)\.
.
.IP
\fBImportant note:\fR If you set the libpath, you must explicitly include the standard library if you want Rosie to include the standard library in its search\. A \fBlibpath\fR set on the command line takes precedence over one set in an initialization file\.
.
.TP
\fB\-\-norcfile\fR
Do not load any initialization file\.
.
.TP
\fB\-o, \-\-output\fR \fIformat\fR
Output match data according to \fBformat\fR, which may be one of:
.
.IP
color, subs, data, json, line
.
.IP
or possibly other (installation\-dependent) values\. Use \fBrosie help\fR to discover what is available in your installation\.
.
.IP
Note the default output style for the \fBmatch\fR command is \fBcolor\fR, and for the \fBgrep\fR command, is \fBline\fR\.
.
.TP
\fB\-\-rcfile\fR \fIfile\fR
Load the initialization file specified, instead of the default, \fB~/\.rosierc\fR\.
.
.TP
\fB\-\-rpl\fR \fIsource\fR
Define additional patterns, import additional packages\. The argument passed to \fB\-\-rpl\fR is typically enclosed in single quotes so that Rosie will see it as a single string with no shell expansions performed\.
.
.TP
\fB\-\-verbose\fR
When appropriate, output additional information\.
.
.SH "COMMAND OPTIONS"
.
.TP
\fB\-h, \-\-help\fR
Display help for a command\. This option must be given after the command name, e\.g\. \fBrosie match \-h\fR\.
.
.TP
\fB\-a, \-\-all\fR
Generate output for all input lines\. Lines that do not match the pattern are written to stderr, while matched output is written to stdout\.
.
.TP
\fB\-w, \-\-wholefile\fR
Match against the whole input file as if it were a single string\.
.
.TP
\fB\-F, \-\-fixed\-strings\fR
Interpret \fIpattern\fR as a set of fixed (literal) strings, instead of an RPL pattern (which reqires double quotes around string literals)\.
.
.TP
\fB\-\fR
Stop reading from the given input files, if any, and start reading from the standard input\.
.
.SH "ENVIRONMENT"
Currently, Rosie does not use any environment variables\. If this would be a useful addition, please open an issue on Rosie\'s GitHub repository to request this enhancement\.
.
.SH "FILES"
.
.TP
\fB~/\.rosierc\fR
This is the default initialization file, which is loaded before any command line options are processed\. The syntax of a Rosie rcfile is essentially the RPL syntax, with two dashes to start comments, and option values in double\-quoted strings\.
.
.IP
Options are set using an assignment\-like syntax, e\.g\.
.
.IP
\fBlibpath = "bar:baz"\fR
.
.IP
to set the rosie \fBlibpath\fR to a sequence of two directories, \fBbar\fR and \fBbaz\fR\.
.
.IP
The recognized options are:
.
.IP
\fBcolors\fR=\fIcolorspec\fR to set \fBcolors\fR to a \fIcolorspec\fR (can be used multiple times; values will be concatenated)
.
.IP
\fBlibpath\fR=\fIpaths\fR to set \fBlibpath\fR (can be used multiple times; values will be concatenated)
.
.IP
\fBloadfile\fR=\fIfile\fR to load the specified file (can be used multiple times to load multiple rpl files)
.
.SH "EXAMPLES"
Forthcoming
.
.SH "BUGS"
Rosie does not normalize Unicode input\. If a pattern contains composed characters, for example, these will not match their decomposed equivalents in the input (and vice versa)\.
.
.P
The RPL compiler is slow (but matching is reasonably fast)\.
.
.P
Issues (bugs and enhancement requests) may be found on the Rosie issue page \fIhttps://github\.com/jamiejennings/rosie\-pattern\-language/issues\fR\.
.
.SH "SEE ALSO"
See Rosie\'s home \fIhttp://rosie\-lang\.org\fR for news, docs, etc\.
.
.SH "AUTHOR"
Jamie A\. Jennings
.
.P
With contributions from the people listed in the CONTRIBUTORS \fIhttps://github\.com/jamiejennings/rosie\-pattern\-language/blob/master/CONTRIBUTORS\fR file\.
.
.P
Email the Rosie Project at \fIinfo@rosie\-lang\.org\fR\.
.
.P
Follow @jamietheriveter \fIhttps://twitter\.com/jamietheriveter\fR on Twitter\.


CHANGELOG
2018-04-24  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-9:

	CHANGED: The output option named "matches" is now named "data" for
	consistency with the others, particularly "subs", since those are both
	components of a match.

	CHANGED: Removed the output option named "default".

	FIXED: Bug in REPL when user enters sample data that is not a valid RPL
	string, e.g. one with an invalid escape sequence.

	FIXED: The pattern "num.float" now accepts all valid JSON numbers.
	Also, "frac" (for "fraction") is now an alias and will no longer appear
	in output from the "num" patterns.

	ENHANCED: When a string containing RPL is loaded into an engine, and
	that string declares that it is a package, say "foo", then a legitimate
	package "foo" is created.  A subsequent call to	engine.load("import foo")
	will now succeed.  An attempt to force a re-import via
	engine.import("foo") will fail, however, because the engine has no way
	of loading the RPL source again (it is not on the file system).

2018-04-23  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-8:

	ADDED: New functions in librosie for clients that want to examine how
	RPL is parsed.  For expressions and blocks (sets of statements), you can
	ask librosie to parse them, compute the free variables (references), and
	compute the packages that need to be imported before the RPL can be
	used.

	CHANGED: It is no longer valid to give an empty string to the load()
	function.  E.g. the CLI will now complain on --rpl ''.

	CHANGED: The net.fqdn pattern used to accept floats like 1.2e03.  It no
	longer does.  You can use net.fdqn_strict to accept all valid partially
	qualified domain names, which would include 1.2e03.

	CHANGED: In rosie.py, the boolean output encoder now returns True or
	False instead of True or None.

	ENHANCED: The cli and librosie now follow symlinks to find the Rosie
	installation files (lib/*.luac, rpl/*.rpl, and others).

	FIXED: The new_engine() function is now thread-safe.  There's a mutex
	around previously unsafe part, so simultaneous engine creation in
	multiple threads is safe, but part of it will only run in one thread at
	a time.

2018-04-10  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-7:

	CHANGED: The rosie.py module for using Rosie with Python now supports a
	load() function.  It is OPTIONAL to use load() in general, but is
	necessary if you have multiple Rosie installations and you want to load
	a specific librosie.so|dylib.

2018-04-08  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-6:

	FIXED: The liblua library now uses a non-locking feof to avoid a
	deadlock that was observed due to a race condition when many threads
	were using rosie.  Related: librosie no longer calls fflush() with a
	NULL argument, which was triggering a walk over all open files.

	FIXED: A build issue related to the (dreaded) use of recursive MAKE.

	CHANGED: A breaking change was made to the Python librosie API in
	rosie.py.  The rosie.engine() function no longer accepts any arguments,
	whereas before it took the path to librosie.  Since the path to find
	librosie can only be set before any engines are created, and will apply
	to all engines, it has its own API: rosie.set_librosie_dir().

	ENHANCED: The multi-threaded test program, src/librosie/C/mt.c, is now
	structured more like a typical use case, in which each thread makes a
	rosie engine, uses it, and then finalizes it.

2018-03-31  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-5:

	ENHANCED: The build process (Makefile) now allows librosie/rosie to be
	built using a relative path to find the rosie installation.  That path
	is relative to the librosie install directory when it starts with a
	double slash ('//'), and relative to the current directory otherwise.
	We recommend AGAINST building with a path that is relative to the
	current directory, for security reasons.

	FIXED: a bug in rosie.py that occured when printing some kinds of
	errors.

	CHANGED: warnings printed by execute_rcfile() are now returned to the
	librosie client instead.

2018-03-16  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-4:

	FIXED: Some commands entered at the CLI did not respect some of the
	command line options, e.g. "test" did not adopt the "--libpath"
	setting.

	FIXED: Some loader/compiler/parser messages, which are called
	"violations", did not convert properly to JSON in librosie for return to
	the calling program.

2018-03-15  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-3:

	ENHANCED: The sample docker files in extra/docker now includes python 3
	and the official arch distro.

	ENHANCED: There should be NO compile warnings from gcc as long as it is
	gcc version 7 or higher.  Older versions will warn about the fallthrough
	attribute, because they don't understand it.

	FIXED: In rosie.py, a compiled pattern is now a proper python object,
	and it contains an engine reference so that the engine cannot be GC'd
	before its compiled patterns.  (This does not happen in Python 2.7, but
	it does in Python 3.6.)

	FIXED: Rosie is once again easy to install on OS X!  See
	https://github.com/jamiejennings/homebrew-rosie

2018-03-09  Jamie Jennings  <jjennings@us.ibm.com>

	* 1.0.0-beta-2:

	CHANGED: The (undocumented) output encoder "none" has been removed.

	NEW: An output encoder called "bool" was added.  It is not useful from
	the command line, but when using librosie's match() function, it will
	return a code 0/1 indicating a match or no match.

	NEW: Python example program extra/examples/generic_sloc.py.

	ENHANCED: When the user specifies an output encoder name that is not
	supported, a proper error is returned.  Previously, the error that was
	returned could be confused with a bug in librosie.

	ENHANCED: `make install` now installs the rosie man page as well.

	ENHANCED: Makefile was tweaked so that it works with the new brew
	(http://brew.sh) formula for install Rosie on OS X.

	FIXED: a bug in pattern net.fqdn_practical that caused it to include a

src/c/rosie.c
/*  -*- Mode: C/l; -*-                                                       */
/*                                                                           */
/* rosie.c      Create a Lua state, load Rosie, expose the Rosie API         */
/*                                                                           */
/*  © Copyright IBM Corporation 2016.                                        */
/*  LICENSE: MIT License (https://opensource.org/licenses/mit-license.html)  */
/*  AUTHOR: Jamie A. Jennings                                                */


/* 
 * Based on src/lua.c from the open source Lua 5.3.2 distribution under MIT license
 *
 * Right now, this is just hacking around to learn my way around using the Lua state.
 *
 */


#define rosie_c

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "lauxlib.h"
#include "lualib.h"

#define PROGNAME		"rosie"

static lua_State *globalL = NULL;
static const char *progname = PROGNAME;

/*
 * Hook set by signal function to stop the interpreter.
 */
static void lstop (lua_State *L, lua_Debug *ar) {
  (void)ar;  /* unused arg. */
  lua_sethook(L, NULL, 0, 0);  /* reset hook */
  luaL_error(L, "interrupted!");
}

/*
** Function to be called at a C signal. Because a C signal cannot
** just change a Lua state (as there is no proper synchronization),
** this function only sets a hook that, when called, will stop the
** interpreter.
*/
static void laction (int i) {
  signal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
  lua_sethook(globalL, lstop, LUA_MASKCALL | LUA_MASKRET | LUA_MASKCOUNT, 1);
}

/*
** Prints an error message, adding the program name in front of it
** (if present)
*/
static void l_message (const char *pname, const char *msg) {
  if (pname) lua_writestringerror("%s: ", pname);
  lua_writestringerror("%s\n", msg);
}

/*
** Check whether 'status' is not OK and, if so, prints the error
** message on the top of the stack. It assumes that the error object
** is a string, as it was either generated by Lua or by 'msghandler'.
*/
static int report (lua_State *L, int status) {
  if (status != LUA_OK) {
    const char *msg = lua_tostring(L, -1);
    l_message(progname, msg);
    lua_pop(L, 1);  /* remove message */
  }
  return status;
}

/*
** Message handler used to run all chunks
*/
static int msghandler (lua_State *L) {
  const char *msg = lua_tostring(L, 1);
  if (msg == NULL) {  /* is error object not a string? */
    if (luaL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
        lua_type(L, -1) == LUA_TSTRING)  /* that produces a string? */
      return 1;  /* that is the message */
    else
      msg = lua_pushfstring(L, "(error object is a %s value)",
                               luaL_typename(L, 1));
  }
  luaL_traceback(L, L, msg, 1);  /* append a standard traceback */
  return 1;  /* return the traceback */
}

/*
** Interface to 'lua_pcall', which sets appropriate message function
** and C-signal handler. Used to run all chunks.
*/
static int docall (lua_State *L, int narg, int nres) {
  int status;
  int base = lua_gettop(L) - narg;  /* function index */
  lua_pushcfunction(L, msghandler);  /* push message handler */
  lua_insert(L, base);  /* put it under function and args */
  globalL = L;  /* to be available to 'laction' */
  signal(SIGINT, laction);  /* set C-signal handler */
  status = lua_pcall(L, narg, nres, base);
  signal(SIGINT, SIG_DFL); /* reset C-signal handler */
  lua_remove(L, base);  /* remove message handler from the stack */
  return status;
}

static int dochunk (lua_State *L, int status) {
  if (status == LUA_OK) status = docall(L, 0, 0);
  return report(L, status);
}

static int dofile (lua_State *L, const char *name) {
  return dochunk(L, luaL_loadfile(L, name));
}

static int dostring (lua_State *L, const char *s, const char *name) {
  return dochunk(L, luaL_loadbuffer(L, s, strlen(s), name));
}

// $SCRIPT_PATH/bin/lua -e "ROSIE_HOME=\"$SCRIPT_PATH\"; SCRIPTNAME=\"$0\"" $SCRIPT_PATH/src/run.lua "$@"

int main (int argc, char **argv) {
  int status;
  lua_State *L = luaL_newstate();  /* create state */
  if (L == NULL) {
    l_message(argv[0], "cannot create state: not enough memory");
    return EXIT_FAILURE;
  }

  luaL_checkversion(L);  /* check that interpreter has correct version ??? */

  luaL_openlibs(L);  /* open standard libraries */

  const char *setup = "ROSIE_HOME=\"/Users/jjennings/Work/Dev/rosie-pattern-language\"; SCRIPTNAME=\"foo\"";
  status = dostring (L, setup, "this is a test");
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;
  
  status = dofile(L, "/Users/jjennings/Work/Dev/rosie-pattern-language/src/bootstrap.lua");
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;

  lua_getglobal(L, "bootstrap");
  lua_insert(L, 1);
  status = lua_pcall(L, 0, 0, 0);
  if (status != LUA_OK)
       l_message(progname, lua_pushfstring(L, "error calling 'bootstrap' (%s)",
					   lua_tostring(L, -1)));
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;
  
  lua_getglobal(L, "require");
  lua_pushstring(L, "repl");
  status = docall(L, 1, 1);  /* call 'require(name)' */
  lua_pop(L, 1);	     /* pop result of require */
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;

  lua_getglobal(L, "require");
  lua_pushstring(L, "api");
  status = docall(L, 1, 1);  /* call 'require(name)' */
  lua_pop(L, 1);	     /* pop result of require */
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;

  lua_getglobal(L, "api");
  lua_getfield(L, -1 , "new_engine");		    /* -1 is stack top, i.e. api table */
  lua_pushstring(L, "REPL ENGINE");
  status = docall(L, 1, 2);                         /* call 'api.new_engine(name)' 1 arg, 2 results */
  report(L, status);
  if (status != LUA_OK) return EXIT_FAILURE;

  lua_pushboolean(L, 1);			    /* push true */
  if (lua_compare(L, -1, -3, LUA_OPEQ) != 1) {
       l_message(progname, lua_pushfstring(L, "error getting engine: %s", lua_tostring(L, -2)));
       return EXIT_FAILURE;
  }

  lua_pop(L, 1);				    /* remove the true we pushed */
  lua_setglobal(L, "eid");			    /* save engine id */ 
  lua_pop(L, 1);				    /* pop the ok returned by new_engine */

  lua_getglobal(L, "repl");	  /* push repl fcn */
  lua_getglobal(L, "eid");	  /* push engine id */
  docall(L, 1, 1);				    /* print(ok) */
  report(L, status);
  if (status != LUA_OK) {
       l_message(progname, lua_pushfstring(L, "error starting repl: %s", lua_tostring(L, -1)));
       return EXIT_FAILURE;
  }

  lua_close(L);
  return (status == LUA_OK) ? EXIT_SUCCESS : EXIT_FAILURE;

}

